### ISSUE TITLE
Performance issues uncovered when training on MGnify (larger dataset)

### ISSUE THREAD

**Issue by idavi-bcs** [Kuhlman-Lab/ThermoMPNN#34]
I recently tried re-training ThermoMPNN on the new MGnify dataset from the Rocklin lab, and uncovered a number of performance-related issues in the code that made it impractical:

- in `protein_mpnn_utils.py`, each call to `parse_PDB_biounits()` reads the PDB file 352 times.  Refactoring this to read the file once and identify the chain(s) first speeds up data loading several fold, and eliminates the need for caching PDB objects as pickles.
- in `datasets.py`, refactoring the Megascale CSV loading to use idiomatic Pandas takes data loading from 2 hours to 2 seconds.
- in `train_thermompnn.py`, calling `self.log()` on every step causes (1) lots of extra traffic between the GPU and main memory and (2) causes an associated O(N^2) slowdown that cripples training when the number of batches becomes much larger than the original ~250.

Additionally, because MGnify has just a few mutations for 100k proteins, instead of hundreds per protein, I find it helps to group multiple proteins into a single batch in order get reasonably stable gradients for training.  I've introduced a `cfg.training.batch_size` parameter, and refactored `shared_eval()` to accept batches larger than 1.  This *seems* to be ok, but I'm not 100% confident this hasn't broken some expectation elsewhere in the code and made my training results worse than they could be.  At least retraining on the original data with batch_size = 1 still seems to work, though.

I would be happy to contribute these changes as a PR, if someone in the lab is available to review/test/merge them?

**Comment by hdieckhaus** [comment]
Hey Ian,

Thanks for your note. You're right, the parsing is not very performant for large datasets. We fixed some of this in [ThermoMPNN-D](https://github.com/Kuhlman-Lab/ThermoMPNN-D), but not all of it. I've been wanting to update ThermoMPNN for some time, probably by wrapping it into the [Foundry](https://github.com/RosettaCommons/foundry/tree/production) framework, I just haven't had the time.

I'd be happy to look over a PR for a new training loop that is better optimized, as long as it doesn't break the original training and inference code. I am curious to see how a retrained model will do with some new data.

Thanks,
Henry

**Comment by idavi-bcs** [comment]
Thanks @hdieckhaus !  Just submitted the PR.
